skrifa/src/variation.rs

```diff
@@ -0,0 +1,539 @@
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use skrifa::prelude::*;
+    /// # fn wrapper(font: &FontRef) {
+    /// let axes = font.axes();
+    /// let mut location = vec![NormalizedCoord::default(); axes.len()];
+    /// axes.location_to_slice(&[("wght", 250.0), ("wdth", 75.0)], &mut location);
+    /// # }
+    /// ```
+    pub fn location_to_slice<I>(&self, settings: I, location: &mut [NormalizedCoord])
+    where
+        I: IntoIterator,
+        I::IntoIter: 'a + Clone,
+        I::Item: Into<VariationSetting>,
+    {
+        for coord in location.iter_mut() {
+            *coord = NormalizedCoord::default();
+        }
+        let avar_mappings = self.avar.as_ref().map(|avar| avar.axis_segment_maps());
+        for setting in settings.into_iter() {
+            let setting = setting.into();
+            // To permit non-linear interpolation, iterate over all axes to ensure we match
+            // multiple axes with the same tag:
+            // https://github.com/PeterConstable/OT_Drafts/blob/master/NLI/UnderstandingNLI.md
+            // We accept quadratic behavior here to avoid dynamic allocation and with the assumption
+            // that fonts contain a relatively small number of axes.
+            for (i, axis) in self
+                .iter()
+                .enumerate()
+                .filter(|v| v.1.tag() == setting.selector)
+            {
+                let coord = axis.record.normalize(Fixed::from_f64(setting.value as f64));
+                let coord = avar_mappings
+                    .as_ref()
+                    .and_then(|mappings| mappings.get(i).transpose().ok())
+                    .flatten()
+                    .map(|mapping| mapping.apply(coord))
+                    .unwrap_or(coord);
+                location[i] = coord.to_f2dot14();
+            }
+        }
+    }
+
+    /// Given an iterator of variation settings in user space, returns a
+    /// new iterator yielding those settings that are valid for this axis
+    /// collection.
+    ///
+    /// * Setting selectors that don't match an axis are dropped.
+    /// * If more than one setting for an axis is provided, the last one is
+    ///     retained.
+    /// * Setting values are clamped to the range of their associated axis.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// # use skrifa::prelude::*;
+    /// # fn wrapper(font: &FontRef) {
+    /// // Assuming a font contains a single "wght" (weight) axis with range
+    /// // 100-900:
+    /// let axes = font.axes();
+    /// let filtered: Vec<_> = axes
+    ///     .filter(&[("wght", 400.0), ("opsz", 100.0), ("wght", 1200.0)])
+    ///     .collect();
+    /// // The first "wght" and "opsz" settings are dropped and the final
+    /// // "wght" axis is clamped to the maximum value of 900.
+    /// assert_eq!(&filtered, &[("wght", 900.0).into()]);
+    /// # }
+    /// ```
+    pub fn filter<I>(&self, settings: I) -> impl Iterator<Item = VariationSetting> + Clone
```

Comment:
This and the previous function are both highly useful for the Skia context. Thanks for adding these.